ARROW-6996: [Python] Expose boolean filter kernel on ChunkedArray/RecordBatch/Table

jorisvandenbossche opened the pull request.

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -344,6 +344,73 @@ cdef class ChunkedArray(_PandasConvertible):
+        #         FilterKernel(_context(), CDatum(self.sp_chunked_array),
+        #                      CDatum(mask.sp_array), &out))
+
+        # return wrap_datum(out)
```
Currently, the C++ Filter version that accepts Datums only accepts Array datums. If that would accept Datums of all kinds (and do the dispatch to the specialized Filter version there), the code here could be simpler (it could be something as the above). 

Would that be a desired feature of the C++ Filter kernel function?

wesm replied:
I brought this up when the ChunkedArray Filter functions were first implemented, the tracking JIRA is https://issues.apache.org/jira/browse/ARROW-7009

jorisvandenbossche replied:
So it's ok to leave that for a separate PR and go ahead here with the different overloads?

wesm commented:
Could this be rehabilitated in time for 0.17.0? Seems useful

jorisvandenbossche commented:
I rebased this, and cleaned up the code (in the meantime, the Filter kernel added Datum capabilities for all types, so that made the implementation here much cleaner)

wesm commented:
Cool, I'm having another look
